### PR TITLE
Added dijkstra version with node weights Fixes #1345

### LINKED ISSUE #1345: Add dijkstra version with node weights
(minor request)

Sometimes people want to find the cheapest path where the cost is calculated based on node weights or a sum of edge and node weights.

In this case it is easy to translate into dijkstra's algorithm with just edge weights by assigning a weight to each edge equal to the sum of the weights of the nodes on either end (or just their average).  However, this requires doing modifications to all the edges a priori, and we may not want to modify the graph.

It should be straightforward to adapt the existing Dijkstra's algorithm to do this at each step as it runs.

### PR DESCRIPTION
**Pull request by MridulS**
@dschult Looking forward to your views and suggestions.

### CODE REVIEW

**Comment by dschult** [comment]
Can you do this without copying the graph? For many graphs copying will require a large amount of memory. (see suggestion at bottom).

The doc_string is not explicit about how weights are computed. And there is more than one way to handle node weights--so you may need to be explicit. The original poster provided two possible node weight methods. How can we best do this? You could ask the poster for the use case to find out which method makes most sense, or whether we need to provide a flexible way to handle it.

You should probably let the user pick different attribute names for node and edge weights.

By using edge[0] and edge[1] you are unpacking the tuple many times. It might be better to unpack it once with ``` for u,v in G.edges(): ``` because then you can use u and v.

Using G[u][v][weight] does not catch errors in a  user friendly way. For example, if an edge does not have the attribute what should we do?

Try copying the dijkstra code and morphing it to allow node weights. It handles special cases so you will see how it does that.  Plus you will get the benefit of learning the algorithm more deeply. :)

**Comment by MridulS** [comment]
@dschult Job 961.10(python pypy)  is failing for some weird reason. In a number of different doctests the output is
```
Expected:
    [1, 2, 3, 4, 5, 6, 7, 8, 9]
Got:
    [8, 9, 1, 7, 6, 5, 4, 3, 2]
```
```
Expected:
    [0, 1, 2, 3, 4, 5, 6]
Got:
    [2, 3, 4, 6, 5, 0, 1]
```
```
Expected:
    [(0, 0), (0, 1), (1, 1), (1, 2), (2, 2)]
Got:
    [(0, 0), (1, 0), (1, 1), (2, 1), (2, 2)]
```
I haven't even made changes in files related to these tests.

**Comment by dschult** [comment]
Those look like sorting issues. The order items are reported from dicts and sets is not specified in Python so different versions and different machines may give different orders. Sounds like this might be another pull request--figure out how other tests have handled potential order differences (search for "sorted" in the tests) and implement them for these tests.  Ah, the joys of testing. :)

**Comment by MridulS** [comment]
@dschult I guess copying the graph is unavoidable, until we change the main dijkstra code to accommodate node weights. I can try to implement dijkstra for a more general case with both weighted nodes and weighted edges. If the user doesn't provides us with those attributes we can make them 0 for respective edges and nodes(as they have to be non-negative numericals).Then we can change the effective weight by the following function 
```
w'(u,v) = w(u,v) + vertex_weight(v)  
```
where ``` w' ``` is the updated weight and then carry on the regular dijkstra algorithm.
But still we need to figure out a way of keeping the graph in its initial state to avoid copying it.

**Comment by dschult** [comment]
The original Dijkstra code does not copy the graph. So start with that code... make it work for node weights.

Your form of adding vertex_weight(v) to the edge weight(u,v) is not the only way to add node weights. See the original post.

**Comment by chebee7i** [comment]
Regarding the failure of tests, if it is just an ordering issue on the results, then yes, you can use sorted(), but you might also consider using graphs with ordered dicts as your test structure.

**Comment by MridulS** [comment]
@dschult found a pretty easy way to generalise dijkstra algorithm. Nodes take a default value of 0 if there is no weight attribute and edges take 1.

**Comment by MridulS** [comment]
@chebee7i In many cases sorted() is used and travis on pypy still failed. Is it something to do with nosetests running on pypy2. And currently the build is failing for networkx master repo, I guess solving that should be the first priority.

**Comment by MridulS** [comment]
After going through http://morepypy.blogspot.in/2015/01/faster-more-memory-efficient-and-more.html and http://bugs.python.org/issue3332 I think this means that dicts that are key-value consistent are not equlivalent due to ordering, and hence the travis fail.

**Comment by chebee7i** [comment]
I can take a deeper look later today.  At first glance, I'm really happy to see those failures :) as it highlights where we have made some bad assumptions about ordering.

Due to our choice of unordered dicts as the default graph class, I'm less enthusiastic about testing doctests now. Many of the examples expect the edges to come out in some particular order, which is not guaranteed. Yes, we can sort the output in a doctest, but then then the doctest is doing non-example-demonstrating things in order to pass the test. In other words, it no longer just a demonstration of the feature---I think this should be the *primary* (and perhaps sole) goal of the doctests. Presently, the doctest has to become convoluted in order to pass. I'd rather doctests use the ordered graph classes, or we don't test docstrings at all anymore. 

Bottom line for me is that making doctests pass comes at a readability cost. Even though it may be minor in some cases, users will still ask:  Why are you sorting the output? And we would have to say: Oh that has nothing to do with the example, we are just doing that so that the output is guaranteed to match what you might see on your end. After which, we should really be asking ourselves: Well, why didn't we just use a data structure that guarantees that fact and hides those details from the user?

I realize this is a bit off topic with your particular failure, but I'm interested in hearing people's opinions.

**Comment by MridulS** [comment]
@chebee7i Opened a new issue. You can post your comment there so that more people can be involved :)

**Comment by MridulS** [comment]
@dschult found a pretty easy way to generalise dijkstra algorithm. Nodes take a default value of 0 if there is no weight attribute and edges take 1.Please have a look. I guess we can ignore pypy travis failure for a while.

**Comment by dschult** [comment]
Can you look at the original posting?   We need to figure out what the appropriate way to handle node weights is.  The PR method is not what was suggested. But what is best? We also need to check for timing impact. This might need to be a separate routine if it slows down the no-node-weight case by a large factor.

**Comment by MridulS** [comment]
According to the original post there are two ways to add node attribute, one that I have implemented and the other in which we add the avg. node weight of u and v in a (u,v) edge. In calculating avg node weight we have to check weight of two nodes while in the other it is done only once, so I guess it is more efficient.

I ran some timeit tests and the changes are quiet visible.
This branch
``` python
In [24]: setup = """import networkx as nx       
   ....: ... G = nx.fast_gnp_random_graph(100000, 0.0001, 0).to_directed()
   ....: ... nx.set_edge_attributes(G, "weight", 1)"""

In [25]: t.timeit('nx.dijkstra_path(G, 1, 1000)', setup=setup, number=10)
Out[25]: 17.272720098495483
```  
master branch
``` python
In [1]: setup = """import networkx as nx
   ...: ... G = nx.fast_gnp_random_graph(100000, 0.0001, 0).to_directed()
   ...: ... nx.set_edge_attributes(G, "weight", 1)"""

In [2]: import timeit as t

In [3]: t.timeit('nx.dijkstra_path(G, 1, 1000)', setup=setup, number=10)
Out[3]: 13.95748782157898
```

I guess we should move this to separate method  ```dijsktra_weighted_nodes_path```.

**Comment by hagberg** [comment]
Yes, I think it should be a separate function.

**Comment by MridulS** [comment]
@hagberg I think if we can compromise on performance, then generalising the dijkstra algorithm for weighted nodes would be better. If you use ```nx.shortest_path``` on a graph with weighted nodes this won't work. Although I have implemented elementary separate functions for weighted nodes, I believe they are not sufficient.

**Comment by MridulS** [comment]
@hagberg How about adding a check in ``` dijkstra_path``` for weighted nodes? If they exist we can run the updated dijkstra algorithm, else continue. This will integrate weighted nodes to other functions in a better way. And there will be considerably less code.

**Comment by MridulS** [comment]
@dschult @hagberg

**Comment by dschult** [comment]
I think we still need different weight names to be possible for the node weight and the edge weight.

Also, this is a large number of new functions. We could just provide the underlying routine(s) and let people make their own version that e.g. only returns the path or only returns the length? Perhaps @hagberg has a feel for that.

To get travis to like it better you could rebase to the current master.

**Comment by MridulS** [comment]
In that case I think we should at least keep ``single_source_dijkstra_weighted_node`` and ``dijkstra_predecessor_and_distance_weighted_node``. 
But still I think these individual functions won't be integrated properly with the rest of code (``nx.shortest_path`` calls for ``nx.single_source_shortest_path``)  How often people encounter weighted nodes? The current dijkstra implementation only considers weighted edges not nodes. ``nx.shortest_path`` won't give the right answer if the graph has weighted nodes.

TODO: Tests and Documentation

**Comment by MridulS** [comment]
@dschult @hagberg

**Comment by MridulS** [comment]
@dschult @hagberg

**Comment by MridulS** [comment]
@dschult @hagberg Any final word on this?

**Comment by dschult** [comment]
Can you make it accept a different attribute name for node_weights than for
edge_weights?

**Comment by MridulS** [comment]
@dschult How about having two different input parameters for weights? `eweight` for edges and `nweight` for nodes.

**Comment by MridulS** [comment]
@dschult @hagberg Thanks to @jfinkels no need to do this.

``` python
In [13]: def func(u, v, d):
....      return G.node[u].get('node_weight', 1)/2 + G.node[v].get('node_weight', 1)/2 + d.get('weight', 1)

In [14]: nx.dijkstra_path(G, 1, 4, weight=func)
Out[14]: [1, 3, 4]
```

This should be enough.
